### PR TITLE
feature: added new cody receipe optimize

### PR DESCRIPTION
**Pull request by Momilijaz96**
## Test plan

Tested the feature locally by running Cody from "Run and Debug" option of VSCode; with different DSA algorithms in python and general DS/ML based applications and projects. The recipe suggests good code optimizationsto improve memory and time optimizations.

### CODE REVIEW

**Comment by Momilijaz96** [comment]
@cla-bot check

**Comment by Momilijaz96** [comment]
@cla-bot check

File: client/cody-shared/src/chat/recipes/vscode-recipes.ts [hunk 1]
```diff
@@ -33,17 +34,18 @@ function init(): void {
         new ChatQuestion(),
         new ExplainCodeDetailed(),
         new ExplainCodeHighLevel(),
-        new InlineChat(),
         new GenerateDocstring(),
         new GenerateTest(),
         new GitHistory(),
         new ImproveVariableNames(),
+        new InlineChat(),
```
Comment (mrnugget) [inline client/cody-shared/src/chat/recipes/vscode-recipes.ts:41]: Why the reorder?

**Review by mrnugget — CHANGES_REQUESTED** [review]
Thanks for the PR, @Momilijaz96! Very interesting. Do you have examples of how it worked? Did you test it on some code and got good results?

There's some linting issues, according to CI:

```


> @ format:check /root/buildkite/build/sourcegraph
--
  | > prettier '**/{*.{js?(on),ts?(x),graphql,md,scss},.*.js?(on)}' --config prettier.config.js --check --write=false
  |  
  | Checking formatting...
  | [warn] client/cody-shared/src/chat/recipes/vscode-recipes.ts
  | [warn] client/cody/src/main.ts
  | [warn] Code style issues found in 2 files. Forgot to run Prettier?

```

Those need to be fixed

**Review by philipp-spiess — COMMENTED** [review]
This is really interesting. I feel like we should have a recipe regarding time and space complexity!! Thank you so much for this.

File: client/cody-shared/src/chat/recipes/vscode-recipes.ts (hunk 1 above)
Comment (Momilijaz96) [reply client/cody-shared/src/chat/recipes/vscode-recipes.ts]: I removed the reordering, I guess that was extra.

File: client/cody-shared/src/chat/recipes/optimize-code.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+        const languageName = getNormalizedLanguageName(selection.fileName)
+        const promptMessage = `Optimise this code in  ${languageName}. \
+            Start your response by telling if the code can/cannot be optimized. \
+            You need to suggest a list of possible optimisation in less than 50 words each,\
+            but if no optimisation is possible just say Code is already optimised. \
+            If the code is optimisable, you provide Big O time/space comparison and \
+            return updated code, however skip these details if any of them is not changed.\
+            For updated code, add inline comments about changes you made and enclose it in triple backticks. \
+            Output format should be: The code can/cannot be optimized. Optimisation Steps: {} Time and Space Usage: {} Updated Code: {} :\n\n\`\`\`${extension}\n${truncatedSelectedText}\n\`\`\`\n${MARKDOWN_FORMAT_PROMPT}`
```
[the remark this answers is not in the document]
Comment (Momilijaz96) [reply client/cody-shared/src/chat/recipes/optimize-code.ts]: I fixed the formatting issues and updated the prompt message to remove spelling inconsistencies.

**Comment by Momilijaz96** [comment]
There are few cases that I could'nt handle, even with iterative prompt development:
a. I could'nt fix the format of the output, I wanted the "optimization steps", "time/space consumption improvement" in text and "updated code block" in a code block, but no changes in the prompt helped me achieve that.
b. I also struggled with the input code not be included in response, but it is included sometimes with current prompt.
c. I tested it on limited file formats and code types.
Let me know what you guys think!
I

**Review by thenamankumar — APPROVED** [review]

**Comment by mrnugget** [comment]
Looking at the examples here (https://github.com/sourcegraph/sourcegraph/pull/51974#issuecomment-1557420056) it seems like the formatting is broken, no? Every reply is prefixed with "Here is the response" and then it's raw Markdown.

That should be fixed before we merge this

File: client/cody-shared/src/chat/recipes/optimize-code.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        const languageName = getNormalizedLanguageName(selection.fileName)
+        const promptMessage = `Optimize the memory and time consumption of this code in  ${languageName}.\
+You first tell if the code can/cannot be optimized, then\
+if the code is optimizable, suggest a numbered list of possible optimizations in less than 50 words each,\
+then provide Big O time/space comparison for old and new code and finally return updated code.\
+Don't include the input code in your response. Beautify the response for better readability.\
+Response format should be: This code can/cannot be optimzed. Optimization Steps: {} Time and Space Usage: {} Updated Code: {}\
+However if no optimization is possible; just say the code is already optimized. \n\n\`\`\`${extension}\n${truncatedSelectedText}\n\`\`\`\n${MARKDOWN_FORMAT_PROMPT}`
+        const assistantResponsePrefix = `Here is the response  \n\`\`\`${extension}\n`
```
Comment (mrnugget) [inline client/cody-shared/src/chat/recipes/optimize-code.ts:37]: I think this breaks the formatting: we don't want every response prefixed with "Here is the response"

**Review by mrnugget — CHANGES_REQUESTED** [review]
See my comment inline and the previous comment further up

File: client/cody-shared/src/chat/recipes/optimize-code.ts (hunk 3 above)
Comment (Momilijaz96) [reply client/cody-shared/src/chat/recipes/optimize-code.ts]: Thanks for pointing that out... removing this response prefix fixed it.
Check out these examples now:
Screenshot 2023-05-23 at 3 39 23 PM

Screenshot 2023-05-23 at 3 39 54 PM
![Screenshot 2023-05-23 at 3 40 18 PM
](https://github.com/sourcegraph/sourcegraph/assets/53872365/5e621c70-e6b7-40dd-beec-830e7f29df06)

**Review by mrnugget — APPROVED** [review]

**Comment by philipp-spiess** [comment]
🎉 

@Momilijaz96 do you mind solving the merge conflict? I can look into it as well if you're busy. Ping me here when you did this so I can run the latest changes on CI and we can ship this 🚀

**Comment by Momilijaz96** [comment]
> 🎉
> 
> @Momilijaz96 do you mind solving the merge conflict? I can look into it as well if you're busy. Ping me here when you did this so I can run the latest changes on CI and we can ship this 🚀

Hey @philipp-spiess I don't have the write access to resolve the conflict.
Screenshot 2023-05-24 at 12 06 29 PM

**Comment by philipp-spiess** [comment]
@Momilijaz96 You should be able to update your forked repository. Try something like this in the terminal. If that doesn't work, I can look into it :)

```
git remote -v # This should list your fork as the `origin` for this to work. 
git fetch
git merge origin/main
```

**Comment by Momilijaz96** [comment]
> 

Alright I guess it's done!

**Comment by Momilijaz96** [comment]
> @Momilijaz96 You should be able to update your forked repository. Try something like this in the terminal. If that doesn't work, I can look into it :)
> 
> ```
> git remote -v # This should list your fork as the `origin` for this to work. 
> git fetch
> git merge origin/main
> ```

Alright, i guess it's ready to be merged.

File: client/cody-shared/src/chat/recipes/vscode-recipes.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import { ChatQuestion } from './chat-question'
```
Comment (philipp-spiess) [inline client/cody-shared/src/chat/recipes/vscode-recipes.ts:1]: Huh there seems to be a conflict here, this file was moved into the `client/cody` folder on `main`

Comment (Momilijaz96) [reply client/cody-shared/src/chat/recipes/vscode-recipes.ts:1]: I accidently merged current branch into main, I "undo" it and pushed again, I guess that would resolve this conflict.

Comment (philipp-spiess) [reply client/cody-shared/src/chat/recipes/vscode-recipes.ts:1]: This file still seems to be added instead. Is it okay if I push onto your branch to try and resolve this??

Comment (Momilijaz96) [reply client/cody-shared/src/chat/recipes/vscode-recipes.ts:1]: yeah sure, go ahead

Comment (Momilijaz96) [reply client/cody-shared/src/chat/recipes/vscode-recipes.ts:1]: Apparently my forked repo's main branch is a few commits ahead of the upstream's main. I am trying to fetch upstream and merge into my fork's main, but that is not introducing any changes.

**Comment by Momilijaz96** [comment]
Hey @philipp-spiess , Thanks for taking care of this. 👍 
Is the conflict resolved now? or something else needs be done from my end.

**Comment by philipp-spiess** [comment]
@Momilijaz96 All done from your end 🎉  I’m just working on getting CI to pass before I can merge this 🙂

**Comment by philipp-spiess** [comment]
@Momilijaz96 There appears to be a bug in our CI setup right now, I've asked internally for some advice. I'll have this PR bookmarked though and will continue to look into it 🙇

**Comment by Momilijaz96** [comment]
> @Momilijaz96 There appears to be a bug in our CI setup right now, I've asked internally for some advice. I'll have this PR bookmarked though and will continue to look into it 🙇

No worries, Sounds good.

**Comment by philipp-spiess** [comment]
Y'all we did it! 🎉  @Momilijaz96 @mrnugget

**Comment by Momilijaz96** [comment]
> Y'all we did it! 🎉 @Momilijaz96 @mrnugget

Great :-)

**Comment by jdorfman** [comment]
This is a fantastic addition, great work @Momilijaz96! 

And thank you @mrnugget @philipp-spiess for reviewing it.

**Pull request merged**